envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotsVersions.kt

```diff
@@ -19,26 +21,54 @@ import java.util.concurrent.ConcurrentHashMap
  * We don't need strong consistency there.
  */
 internal class SnapshotsVersions {
+    companion object {
+        fun newVersion(): String = UUID.randomUUID().toString().replace("-", "")
+    }
 
     private val versions = ConcurrentHashMap<Group, VersionsWithData>()
 
-    fun version(group: Group, clusters: List<Cluster>, endpoints: List<ClusterLoadAssignment>): Version {
+    fun version(
+        group: Group,
+        clusters: List<Cluster>,
+        endpoints: List<ClusterLoadAssignment>,
+        listeners: List<Listener> = listOf()
+    ): Version {
         val versionsWithData = versions.compute(group) { _, previous ->
             val version = when (previous) {
-                null -> Version(clusters = ClustersVersion(newVersion()), endpoints = EndpointsVersion(newVersion()))
+                null -> Version(
+                        clusters = ClustersVersion(clusters),
+                        endpoints = EndpointsVersion(clusters),
+                        listeners = ListenersVersion(newVersion()),
+                        routes = RoutesVersion(newVersion())
+                )
                 else -> {
                     val clustersChanged = previous.clusters != clusters
+                    val listenersChanged = previous.listeners != listeners
                     Version(
-                        clusters = selectClusters(previous, clustersChanged),
-                        endpoints = selectEndpoints(previous, endpoints, clustersChanged)
+                        clusters = selectClusters(previous, clusters, clustersChanged),
+                        endpoints = selectEndpoints(previous, endpoints, clustersChanged),
+                        listeners = selectListeners(previous, listenersChanged),
+                        routes = selectRoutes(previous, listenersChanged, clustersChanged)
                     )
                 }
             }
-            VersionsWithData(version, clusters, endpoints)
+            VersionsWithData(version, clusters, endpoints, listeners)
         }
         return versionsWithData!!.version
     }
 
+    private fun selectRoutes(
+        previous: VersionsWithData,
+        listenersChanged: Boolean,
+        clustersChanged: Boolean
+    ): RoutesVersion {
+        return if (listenersChanged || clustersChanged) RoutesVersion(newVersion()) else previous.version.routes
```

Comment:
@MarcinFalkowski @lukidzi - you should've stopped me there 😉

Reply:
Look here, I think it was deliberate: https://github.com/allegro/envoy-control/pull/94#discussion_r396331292